// @langchain/mongodb/dist/vectorstores.js
import { VectorStore } from "@langchain/core/vectorstores";
import { chunkArray } from "@langchain/core/utils/chunk_array";
import { Document } from "@langchain/core/documents";
import { maximalMarginalRelevance } from "@langchain/core/utils/math";
import { AsyncCaller } from "@langchain/core/utils/async_caller";
//#region src/vectorstores.ts
/**
* Stub embeddings for auto-embedding mode.
* Throws if any embedding method is called, since the server handles embeddings.
*/
var AutoEmbeddingStub = class {
	async embedDocuments(_texts) {
		throw new Error("Embeddings not available when using auto-embedding mode. The MongoDB Atlas server handles all embedding generation.");
	}
	async embedQuery(_text) {
		throw new Error("Embeddings not available when using auto-embedding mode. The MongoDB Atlas server handles all embedding generation.");
	}
};
/**
* Class that is a wrapper around MongoDB Atlas Vector Search. It is used
* to store embeddings in MongoDB documents, create a vector search index,
* and perform K-Nearest Neighbors (KNN) search with an approximate
* nearest neighbor algorithm.
*/
var MongoDBAtlasVectorSearch = class MongoDBAtlasVectorSearch extends VectorStore {
	collection;
	indexName;
	textKey;
	embeddingKey;
	primaryKey;
	caller;
	useAutoEmbedding;
	_vectorstoreType() {
		return "mongodb_atlas";
	}
	constructor(embeddingsOrDbConfig, args) {
		let embeddings;
		let libArgs;
		let useAutoEmbedding = false;
		if (args !== void 0) {
			embeddings = embeddingsOrDbConfig;
			libArgs = args;
		} else {
			libArgs = embeddingsOrDbConfig;
			embeddings = new AutoEmbeddingStub();
			useAutoEmbedding = true;
		}
		super(embeddings, libArgs);
		this.collection = libArgs.collection;
		this.indexName = libArgs.indexName ?? "default";
		this.textKey = libArgs.textKey ?? "text";
		this.embeddingKey = libArgs.embeddingKey ?? "embedding";
		this.primaryKey = libArgs.primaryKey ?? "_id";
		this.caller = new AsyncCaller(libArgs);
		this.useAutoEmbedding = useAutoEmbedding;
		this.collection.db.client.appendMetadata({ name: "langchainjs_vector" });
	}
	/**
	* Method to add vectors and their corresponding documents to the MongoDB
	* collection.
	* @param vectors Vectors to be added.
	* @param documents Corresponding documents to be added.
	* @returns Promise that resolves when the vectors and documents have been added.
	*/
	async addVectors(vectors, documents, options) {
		if (this.useAutoEmbedding) throw new Error("Cannot add vectors directly when using auto-embedding mode.");
		const docs = vectors.map((embedding, idx) => ({
			[this.textKey]: documents[idx].pageContent,
			[this.embeddingKey]: embedding,
			...documents[idx].metadata
		}));
		if (options?.ids === void 0) await this.collection.insertMany(docs);
		else {
			if (options.ids.length !== vectors.length) throw new Error(`If provided, "options.ids" must be an array with the same length as "vectors".`);
			const { ids } = options;
			for (let i = 0; i < docs.length; i += 1) await this.caller.call(async () => {
				await this.collection.updateOne({ [this.primaryKey]: ids[i] }, { $set: {
					[this.primaryKey]: ids[i],
					...docs[i]
				} }, { upsert: true });
			});
		}
		return options?.ids ?? docs.map((doc) => doc[this.primaryKey]);
	}
	/**
	* Method to add documents to the MongoDB collection.
	*
	* In manual embedding mode: converts documents to vectors using embeddings, then inserts.
	* In auto-embed mode: inserts documents with text only; MongoDB server handles embedding.
	*
	* @param documents Documents to be added.
	* @returns Promise that resolves when the documents have been added.
	*/
	async addDocuments(documents, options) {
		if (this.useAutoEmbedding) {
			const docs = documents.map((document) => ({
				[this.textKey]: document.pageContent,
				...document.metadata
			}));
			if (options?.ids === void 0) await this.collection.insertMany(docs);
			else {
				if (options.ids.length !== documents.length) throw new Error(`If provided, "options.ids" must be an array with the same length as "documents".`);
				const { ids } = options;
				for (let i = 0; i < docs.length; i += 1) await this.caller.call(async () => {
					await this.collection.updateOne({ [this.primaryKey]: ids[i] }, { $set: {
						[this.primaryKey]: ids[i],
						...docs[i]
					} }, { upsert: true });
				});
			}
			return options?.ids ?? docs.map((doc) => doc[this.primaryKey]);
		} else {
			const texts = documents.map(({ pageContent }) => pageContent);
			return this.addVectors(await this.embeddings.embedDocuments(texts), documents, options);
		}
	}
	/**
	* Method that performs a similarity search on the vectors stored in the
	* MongoDB collection. It returns a list of documents and their
	* corresponding similarity scores.
	* @param query Query vector for the similarity search.
	* @param k Number of nearest neighbors to return.
	* @param filter Optional filter to be applied.
	* @returns Promise that resolves to a list of documents and their corresponding similarity scores.
	*/
	async similaritySearchVectorWithScore(query, k, filter) {
		if (this.useAutoEmbedding) throw new Error("Cannot perform similarity search with vectors directly when using auto-embedding mode.");
		const postFilterPipeline = filter?.postFilterPipeline ?? [];
		const preFilter = filter?.preFilter || filter?.postFilterPipeline || filter?.includeEmbeddings ? filter.preFilter : filter;
		const removeEmbeddingsPipeline = !filter?.includeEmbeddings ? [{ $project: { [this.embeddingKey]: 0 } }] : [];
		const pipeline = [
			{ $vectorSearch: {
				queryVector: MongoDBAtlasVectorSearch.fixArrayPrecision(query),
				index: this.indexName,
				path: this.embeddingKey,
				limit: k,
				numCandidates: 10 * k,
				...preFilter && { filter: preFilter }
			} },
			{ $set: { score: { $meta: "vectorSearchScore" } } },
			...removeEmbeddingsPipeline,
			...postFilterPipeline
		];
		return this.collection.aggregate(pipeline).map((result) => {
			const { score, [this.textKey]: text, ...metadata } = result;
			return [new Document({
				pageContent: text,
				metadata
			}), score];
		}).toArray();
	}
	/**
	* Performs similarity search using text-based queries (auto-embedding mode) or vector queries (manual embedding mode).
	* In auto-embed mode, the text query is sent to MongoDB which handles embedding server-side.
	* In manual embedding mode, the text is embedded client-side and passed as a vector.
	*
	* @param query - Text query for finding similar documents.
	* @param k - Number of similar results to return. Defaults to 4.
	* @param filter - Optional filter based on `FilterType`.
	* @returns A promise resolving to an array of tuples, each containing a
	*          document and its similarity score.
	*/
	async similaritySearchWithScore(query, k = 4, filter = void 0) {
		if (this.useAutoEmbedding) return this.textBasedSearchWithScore(query, k, filter);
		const queryEmbedding = await this.embeddings.embedQuery(query);
		return this.similaritySearchVectorWithScore(queryEmbedding, k, filter);
	}
	/**
	* Searches for documents similar to a text query by embedding the query and
	* performing a similarity search on the resulting vector.
	*
	* @param query - Text query for finding similar documents.
	* @param k - Number of similar results to return. Defaults to 4.
	* @param filter - Optional filter based on `FilterType`.
	* @returns A promise resolving to an array of `DocumentInterface` instances representing similar documents.
	*/
	async similaritySearch(query, k = 4, filter = void 0) {
		return (await this.similaritySearchWithScore(query, k ?? 4, filter)).map(([doc]) => doc);
	}
	async textBasedSearchWithScore(query, k, filter) {
		const postFilterPipeline = filter?.postFilterPipeline ?? [];
		const preFilter = filter?.preFilter || filter?.postFilterPipeline || filter?.includeEmbeddings ? filter.preFilter : filter;
		const removeEmbeddingsPipeline = !filter?.includeEmbeddings ? [{ $project: { [this.embeddingKey]: 0 } }] : [];
		const pipeline = [
			{ $vectorSearch: {
				query: { text: query },
				index: this.indexName,
				path: this.textKey,
				limit: k,
				numCandidates: 10 * k,
				...preFilter && { filter: preFilter }
			} },
			{ $set: { score: { $meta: "vectorSearchScore" } } },
			...removeEmbeddingsPipeline,
			...postFilterPipeline
		];
		return this.collection.aggregate(pipeline).map((result) => {
			const { score, [this.textKey]: text, ...metadata } = result;
			return [new Document({
				pageContent: text,
				metadata
			}), score];
		}).toArray();
	}
	/**
	* Return documents selected using the maximal marginal relevance.
	* Maximal marginal relevance optimizes for similarity to the query AND diversity
	* among selected documents.
	* Not supported in auto-embedding mode.
	* @param {string} query - Text to look up documents similar to.
	* @param {number} options.k - Number of documents to return.
	* @param {number} options.fetchK=20 - Number of documents to fetch before passing to the MMR algorithm (manual embedding mode only).
	* @param {number} options.lambda=0.5 - Number between 0 and 1 that determines the degree of diversity among the results,
	*                 where 0 corresponds to maximum diversity and 1 to minimum diversity.
	* @param {MongoDBAtlasFilter} options.filter - Optional Atlas Search operator to pre-filter on document fields
	*                                      or post-filter following the search.
	*
	* @returns {Promise<Document[]>} - List of documents selected by maximal marginal relevance.
	*/
	async maxMarginalRelevanceSearch(query, options) {
		if (this.useAutoEmbedding) throw new Error("Cannot perform MMR search with vectors directly when using auto-embedding mode.");
		const { k, fetchK = 20, lambda = .5, filter } = options;
		const queryEmbedding = await this.embeddings.embedQuery(query);
		const includeEmbeddingsFlag = options.filter?.includeEmbeddings || false;
		const includeEmbeddingsFilter = {
			...filter,
			includeEmbeddings: true
		};
		const resultDocs = await this.similaritySearchVectorWithScore(MongoDBAtlasVectorSearch.fixArrayPrecision(queryEmbedding), fetchK, includeEmbeddingsFilter);
		return maximalMarginalRelevance(queryEmbedding, resultDocs.map((doc) => doc[0].metadata[this.embeddingKey]), lambda, k).map((idx) => {
			const doc = resultDocs[idx][0];
			if (!includeEmbeddingsFlag) delete doc.metadata[this.embeddingKey];
			return doc;
		});
	}
	/**
	* Delete documents from the collection
	* @param ids - An array of document IDs to be deleted from the collection.
	*
	* @returns - A promise that resolves when all documents deleted
	*/
	async delete(params) {
		const chunkIds = chunkArray(params.ids, 50);
		for (const chunk of chunkIds) await this.collection.deleteMany({ _id: { $in: chunk } });
	}
	static async fromTexts(texts, metadatas, embeddingsOrDbConfig, dbConfig) {
		const docs = [];
		for (let i = 0; i < texts.length; i += 1) {
			const metadata = Array.isArray(metadatas) ? metadatas[i] : metadatas;
			const newDoc = new Document({
				pageContent: texts[i],
				metadata
			});
			docs.push(newDoc);
		}
		if (dbConfig !== void 0) return this.fromDocuments(docs, embeddingsOrDbConfig, dbConfig);
		else return this.fromDocuments(docs, embeddingsOrDbConfig);
	}
	static async fromDocuments(docs, embeddingsOrDbConfig, dbConfig) {
		let embeddings;
		let finalDbConfig;
		let instance;
		if (dbConfig !== void 0) {
			embeddings = embeddingsOrDbConfig;
			finalDbConfig = dbConfig;
			instance = new this(embeddings, finalDbConfig);
		} else {
			finalDbConfig = embeddingsOrDbConfig;
			instance = new this(finalDbConfig);
		}
		await instance.addDocuments(docs, { ids: finalDbConfig.ids });
		return instance;
	}
	/**
	* Static method to fix the precision of the array that ensures that
	* every number in this array is always float when casted to other types.
	* This is needed since MongoDB Atlas Vector Search does not cast integer
	* inside vector search to float automatically.
	* This method shall introduce a hint of error but should be safe to use
	* since introduced error is very small, only applies to integer numbers
	* returned by embeddings, and most embeddings shall not have precision
	* as high as 15 decimal places.
	* @param array Array of number to be fixed.
	* @returns
	*/
	static fixArrayPrecision(array) {
		return array.map((value) => {
			if (Number.isInteger(value)) return value + 1e-15;
			return value;
		});
	}
};
//#endregion
export { MongoDBAtlasVectorSearch };
